File: PHP/AZodiacCalc/AZodiacCalc.php
<?php
/**
 * Created by PhpStorm.
 * User: Rich
 * Date: 11/4/14
 * Time: 12:00 AM
 */
function findSign($_birthdate) {     //looks up asian zodiac sign in Element Tree using birth date
    global $signLookup;
    $_birthyear = date('Y',$_birthdate);    //get the year from the birth date
    $_yearnode = $signLookup->xpath("year/FromYear[.=$_birthyear]/parent::*"); //Find Record for that year
    $_prevYearnode = $signLookup->xpath("year/FromYear[.=$_birthyear-1]/parent::*"); //Find Record for prev. year
    $_fromday = ($_yearnode[0]->FromDay)+1-1;                       //get day of chinese new year
    $_chnewyear = mktime(null,null,null,2,$_fromday,$_birthyear);   //create date for Chinese New year
    if ($_birthdate >= $_chnewyear) {
        $_animal = $_yearnode[0]->Sign;
    } else {
        $_animal = $_prevYearnode[0]->Sign;
    }
    return $_animal;
}

function displayTraits($_animal) {                                  //output Traits based on the sign
    global $traitLookup;
                                                                    //Get record for sign passed to it
    $_signNode = $traitLookup->xpath("sign/Animal[. ="."'".$_animal."'"."]/parent::*");
                                                                    // blasted Regex can't live with it,
                                                                    // can't live without it!
    $_keyword = $_signNode[0]->Keyword;
    $_detail = $_signNode[0]->Attributes;
    $_compatible = $_signNode[0]->MostCompatible;
    $_incompatible = $_signNode[0]->LeastCompatible;
    print "\nYour Asian zodiac sign is: " . $_animal. "<br>";
    print "The " . $_animal . " sign's major trait is " ?>  <b> <?php echo $_keyword ;?> </b>

    <div style="width: 66%">
        <p><?php echo $_detail . "<br>" ?></p>
        <p>
            <?php
            echo "People born under the sign of the ".$_animal. " are most compatible with the signs " . $_compatible;
            echo ", and are least compatible with the ". $_incompatible . " Sign.<br>"
            ?>
        </p>
    </div>
<?php
}
// Main Program
                                                                    //Get input from form
$BirthDate=mktime(null,null,null,$_POST['BirthMonth'],$_POST['BirthDay'],$_POST['BirthYear']);
echo "you input: " . date("m-d-Y",$BirthDate) ."<br><br>";
                                                                    //Read XML Files into Element Trees
$signLookup=simplexml_load_file("AnimalSigns.xml") or die("Error: Cannot Create object");
$traitLookup=simplexml_load_file("AsianZodiac.xml") or die("Error: Cannot Create object");

$Animal = findSign($BirthDate)."";                                  //findSign function returns animal name

displayTraits($Animal);                                             //displayTraits function outputs results
?>

File: python/A_ZodiacCalc.py
#! /usr/bin/python
# A_ZodiacCalc.py
# Richard Forman
# Script to calculate a Chinese Horoscope sign from
# a Birthday (in digits with no leading zeros) and
# output a text file with the zodiac sign's details.
# Two xml files, AsianZodiac.xml and AnimalSigns.xml,
# are included in the zip and should uncompress
# in the same folder as the script.
# Also this program requires the xml.etree.ElementTree
# module to be installed.

usage = "%prog [options] $BirthMonth $BirthDay $BirthYear"
# i.e. AsianZodiacCalc.py 12 25 1999

import sys, optparse, textwrap, time
import xml.etree.ElementTree as ET
from datetime import date
from xml.etree.ElementTree import tostring, XMLID, dump

Chn_MinYear = 1936
Chn_MaxYear = 2032
in_error = "invalid command line arguments to program.\
 Please supply Date between ", Chn_MinYear," and ", Chn_MaxYear," as m d yyyy "

p = optparse.OptionParser(usage=usage)      # create object for parsing input
opts, args = p.parse_args()

for i in range(len(args)):			# convert args from strings to int
	args[i]=int(args[i])
									# Check the arguments
if ((len(args) < 3) or (args[2] < Chn_MinYear) or (args[2] > Chn_MaxYear)):	
	p.error (in_error)				# exit if not enough arguments or year out of range
	                                # create date object for birth date
b_date = date(args[2], args[0], args[1])

									# read Chinese Calendar years into tree
A_Signs = ET.parse('AnimalSigns_attrib.xml')
years = A_Signs.getroot()

									# read Chinese Zodiac Descriptions into tree
A_Zodiac = ET.parse('AsianZodiac_ID.xml')
animals = A_Zodiac.getroot()
Exported = tostring(animals)		# Convert to string so XMLID can create a ID Map
animals_id, id_map = XMLID(Exported)

									# create date object for Chinese new year for that birth year							
year = years.findall("*[@FromYear='"+str(b_date.year)+"']") 
CNY_date = date(b_date.year,2,int(year[0].get('FromDay')))
								
if (b_date >= CNY_date):			# if birthdate is on or after chinese newyear
	animal = year[0].get('Sign')	# animal = sign for that year	
else:								# else Animal is sign for previous year
	year = years.findall("*[@FromYear='"+str(b_date.year-1)+"']")
	animal = year[0].get('Sign')

traits = id_map.get(animal)			# Lookup Animal Traits in Database

print "\nIf your birth date is", b_date.strftime("%B %d %Y,")\
 ,"your Asian Zodiac sign is", animal
print "The",animal,"sign's main trait is",traits.find('Keyword').text
print "\nSee the file animal.txt for more info.\n\n";

original = sys.stdout				# placeholder for console output
myout = open('animal.txt','w')		
sys.stdout = myout					# redirect output to animal.txt file

print "\n"+textwrap.fill(traits.find('Attributes').text).strip()+"\n"
print "You are most compatible with the signs", traits.find('MostCompatible').text+"\n"
print "You are least compatible with the", traits.find('LeastCompatible').text,"sign."+"\n"

sys.stdout = original				# restore output to console 
myout.close							# close file